crates/deltalake-core/src/operations/constraints.rs

```diff
@@ -0,0 +1,263 @@
+
+        Box::pin(async move {
+            if this.name.is_none() {
+                return Err(DeltaTableError::Generic("No name provided".to_string()));
+            } else if this.expr.is_none() {
+                return Err(DeltaTableError::Generic(
+                    "No expression provided".to_string(),
+                ));
+            }
+            let name = this.name.unwrap();
+            let expr = match this.expr {
+                Some(Expression::String(s)) => s,
+                Some(Expression::DataFusion(e)) => e.to_string(),
+                None => unreachable!(),
+            };
+
+            let mut metadata = this
+                .snapshot
+                .metadata()
+                .ok_or(DeltaTableError::NoMetadata)?
+                .clone();
+            let configuration_key = format!("delta.constraints.{}", name);
+
+            if metadata.configuration.contains_key(&configuration_key) {
+                return Err(DeltaTableError::Generic(format!(
+                    "Constraint with name: {} already exists, expr: {}",
+                    name, expr
+                )));
+            }
+
+            let state = this.state.unwrap_or_else(|| {
+                let session = SessionContext::new();
+                register_store(this.log_store.clone(), session.runtime_env());
+                session.state()
+            });
+
+            // Checker built here with the one time constraint to check.
+            let checker = DeltaDataChecker::with_constraints(vec![Constraint::new("*", &expr)]);
+
+            let files_to_check =
+                find_files(&this.snapshot, this.log_store.clone(), &state, None).await?;
+            let scan = DeltaScanBuilder::new(&this.snapshot, this.log_store.clone(), &state)
+                .with_files(&files_to_check.candidates)
+                .build()
+                .await?;
+
+            let task_ctx = Arc::new(TaskContext::from(&state));
+            let record_stream: SendableRecordBatchStream = scan.execute(0, task_ctx)?;
+            let records = collect_sendable_stream(record_stream).await?;
+
+            for batch in records {
+                checker.check_batch(&batch).await?;
+            }
+
+            // We have validated the table passes it's constraints, now to add the constraint to
+            // the table.
+
+            metadata
+                .configuration
+                .insert(format!("delta.constraints.{}", name), Some(expr.clone()));
+
+            let old_protocol = this.snapshot.protocol();
+            let protocol = Protocol {
+                min_reader_version: if old_protocol.min_reader_version > 1 {
+                    old_protocol.min_reader_version
+                } else {
+                    1
+                },
+                min_writer_version: if old_protocol.min_reader_version > 3 {
+                    old_protocol.min_reader_version
+                } else {
+                    3
+                },
```

Comment:
We have some mix-ups here when to use reader and writer version?

Also maybe add a comment that this would be the least required versions for the check feature.

Reply:
Done and done